### ISSUE TITLE
Allow external executor pool for IndexWriter

### ISSUE THREAD

**Issue by raultang** [quickwit-oss/tantivy#858]
**Is your feature request related to a problem? Please describe.**
While having lot of index would create many threads.

**Describe the solution you'd like**
Allow pass executor as params for IndexWriter & IndexReader/Searcher, there is mention for searcher https://github.com/tantivy-search/tantivy/issues/702
`prepare_commit` will wait the previous thread finish and then recreate thread, which might can replace with Sender/Receiver if we use executor for IndexWriter.
The near realtime indexing feature which could change a lot for index commit, so this could consider during that time I think.

**Comment by fulmicoton** [comment]
This is an interesting feature, but a proper solution will require a bit of work. I'll call the user need "multi tenancy".

It has many different solution with very different trade-offs.

Sharing thread is nice, but you might benefit a lot from sharing the buffers as well, 
which could be possible if commit can be synced.

In some use case it is also possible to put everything into one big multitenant index and use one field to describe the 
index considered. The DocID could be sorted by tenant... A special datastructure could hold the boundaries of each tenant.

Could you describe your particular use case? Videoconference or email is ok too.

**Comment by raultang** [comment]
I understand there might have a bit of work of it especially how the buffer allocate for each index. I like your multi tenant proposal.

My use case is more likely in distributed search engine, and there are bunch of index, each instance need maintain many of those, say 100 index for each instance,  and I would want use as many as cpu core for each index, so if my cpu core is 24, so in this situation will create 2400 thread, soon the heap will be eat up as the stack size/buffer if we keep creating index. And to create many of thread of this might lead other performance issue like context switch etc.

**Comment by fulmicoton** [comment]
> this situation will create 2400 thread, soon the heap will be eat up as the stack size/buffer if we keep creating index

The indexing buffer is dwarfing the thread stacks here.

>  I would want use as many as cpu core for each index, so if my cpu core is 24

Why would you want that? 
The stream of docs being indexing is very unstable and your might have 100% of docs being indexed on index 1 for 1 mn and then 100% on index 2 or something like that?

**Comment by raultang** [comment]
> The indexing buffer is dwarfing the thread stacks here.

True, I meant it was using many of the resource somehow will reach the limit.

> Why would you want that?

I want to index docs as soon as possible for each index, which can get the maximum performance, because most of the time not all indices need to be doc.

> The stream of docs being indexing is very unstable and your might have 100% of docs being indexed on index 1 for 1 mn and then 100% on index 2 or something like that?

So let all indices are using same buffer and to compete the buffer resource might can control the overall injection.
